### PR TITLE
Require Functor over Bifunctor where possible

### PR DESCRIPTION
**Pull request by expipiplus1**
Fixes https://github.com/ekmett/bifunctors/issues/124

   Note that this does add a Functor requirement where none was required
   before

### CODE REVIEW

**Comment by expipiplus1** [comment]
Rebased onto 5.15: https://gist.github.com/expipiplus1/4cc5e35895e0559487835032e34d8e88

File: src/Data/Bifunctor/TH.hs [hunk 1]
```diff
@@ -1122,9 +1124,12 @@ functorLikeTraverse tvMap FT { ft_triv = caseTrivial,     ft_var = caseVar
...
       (xrs, xcs) <- unzip <$> mapM (go co) args
-      let numLastArgs, numFirstArgs :: Int
-          numLastArgs  = min 2 $ length args
-          numFirstArgs = length args - numLastArgs
+      let numFirstArgs :: Int
+          mentionsMappedVar :: Type -> Bool
+          mentionsMappedVar = any (`Map.member` tvMap) . freeVariables
+          (firstArgs, lastArgs) = break mentionsMappedVar args
+          numFirstArgs = length firstArgs
+          numLastArgs = length lastArgs
```
Comment (RyanGlScott) [inline src/Data/Bifunctor/TH.hs:1132]: This is performing redundant work, as `go` is already computing whether each argument to the type constructor contains occurrences of the last two type parameters. A more efficient way to implement this would be something like this:

```hs
      let (xcFirsts, xcLasts) = break id xcs
          numLastArgs, numFirstArgs :: Int
          numFirstArgs = length xcFirsts
          numLastArgs = length xcLasts
```

File: src/Data/Bifunctor/TH.hs [hunk 2]
```diff
@@ -1143,6 +1148,8 @@ functorLikeTraverse tvMap FT { ft_triv = caseTrivial,     ft_var = caseVar
           -> tuple $ Unboxed len
           |  fc || or (take numFirstArgs xcs)
           -> wrongArg                    -- T (..var..)    ty_1 ... ty_n
+          | numLastArgs > 2
+          -> outOfPlaceTyVarError conName
```
Comment (RyanGlScott) [inline src/Data/Bifunctor/TH.hs:1152]: By definition, none of the `numFirstArgs` elements of `xcs` contain occurrences of the last two type parameters, so we no longer need to check `or (take numFirstArgs xcs)` here. I'd propose writing this like so:

```hs
          |  fc || numLastArgs > 2
          -> wrongArg                    -- T (..var..)    ty_1 ... ty_n
```

Moreover, you don't need to call `outOfPlaceTyVarError conName`, as `wrongArg` is meant to abstract over how each derived class chooses to deal with occurrences of out-of-place type variables. (This avoids having to pass `conName` to `functorLikeTraverse` in the first place.)

File: src/Data/Bifunctor/TH.hs
Comment (RyanGlScott) [inline src/Data/Bifunctor/TH.hs:1]: Do make sure to update the Haddocks [here](https://github.com/ekmett/bifunctors/blob/ca1116599b85b2344118aed2e49ebb707142598a/src/Data/Bifunctor/TH.hs#L127-L131) to describe the new behavior implemented in this PR.

**Review by RyanGlScott — CHANGES_REQUESTED** [review]

**Comment by expipiplus1** [comment]
Thank you for the patient review, @RyanGlScott!

Still to fix is the instance context generation, as it still wants to generate `Bifunctor` instances for any type variable `f : * -> * -> *`. This should now depend on its usage.

For example this, although `f` and `g` both have kind `* -> * -> *`, `f` requires a `Bifunctor` instance and `g` a `Functor` instance:

```haskell
data Baz f g a b where
  Baz1 :: f a b -> Baz f g a b
  Baz2 :: g Int b -> Baz f g a b

deriveBifunctor ''Baz
```

An alternative would be to not change this context generation heuristic and simply use the old behavior of requiring a `Bifunctor` instance when dealing with fields with two type variables (regardless of if they're ours). (Nicest yet would be just put a wildcard for the instance context, and allow GHC to fill it with the inferred context (which is sadly not permitted today))

**Comment by expipiplus1** [comment]
Another alternative would be to allow configuring this behavior via `Options`, something like `preferFmap`, that way people who can tolerate writing their own instance head can use that and allow type variables with arity > 2 to still use `Functor` where possible.

File: src/Data/Bifunctor/TH.hs
Comment (expipiplus1) [reply src/Data/Bifunctor/TH.hs:1]: As described below, I've ended up not changing this behaviour

**Comment by RyanGlScott** [comment]
Ugh, I completely forgot about instance context generation. I'll be the first person to admit that the way `bifunctors` infers instance contexts is imperfect, and that's largely to be expected. Doing It Right™ would require something akin to full-blown type inference, and that is so difficult to do with Template Haskell that I don't think it is worth bothering to try.

This is all to say: I'm not terribly bothered by the fact that `deriveBifunctor ''Baz` would attempt to generate `instance (Bifunctor f, Bifunctor g) => Bifunctor (Baz f g)`. Examples like `Baz2 :: g Int b -> Baz f g a b` are not the common case, so I'm fine with making users manually write the instance context out for such instances (in conjunction with `makeBimap`). Moreover, this generated instance will actually _typecheck_ with GHC 9.6 due to `Bifunctor` having a quantified `Functor` superclass, so this bothers me even less.

What _does_ bother me a bit is the fact that these two instances:

```hs
data Scooby1 f a b = Scooby1 (f Int a)
data Scooby2 a b = Scooby2 (Either Int a)
```

Will interact quite differently with `deriveBifunctor`. The former would use `bimap` in its generated instance, while the latter will use `fmap`. I'm not a fan of this because I generally like things to be invariant under substitution, and `Scooby2` is basically the same thing as `Scooby1`, where `f` is instantiated to `Either`. Ideally, calling `bimap` on `Scooby1 Either` would behave exactly the same as calling `bimap` on `Scooby2`, but that isn't the case with the way the patch is set up currently.

To summarize, I'd prefer using `fmap` uniformly for things that look like `F Int a` or `f Int a`, even if that doesn't always work with `bifunctors`' limited type inference capabilities. You could also guard this functionality behind `Options` if you'd like—I don't have a strong opinion on that one.
